cron.go

```diff
@@ -170,129 +192,220 @@ func New(opts ...CronOpt) (*Cron, error) {
-	for _, entry := range c.entries {
-		if entry.Job.Name == jobName {
-			found = true
-			continue
-		}
-		// Keep the entries that don't match the specified jobName
-		updatedEntries = append(updatedEntries, entry)
-	}
-	if !found {
-		return fmt.Errorf("job not found: %s", jobName)
+	entry, ok := c.entries[jobName]
+	if !ok || (entry == nil) {
+		return nil
 	}
-	c.entries = updatedEntries
-	return nil
+
+	return &entry.Job
 }
 
 func (c *Cron) ListJobsByPrefix(prefix string) []*Job {
-	c.entriesMutex.RLock()
-	defer c.entriesMutex.RUnlock()
-
 	var appJobs []*Job
+	c.operationsMutex.RLock()
 	for _, entry := range c.entries {
 		if strings.HasPrefix(entry.Job.Name, prefix) {
 			// Job belongs to the specified app_id
 			appJobs = append(appJobs, &entry.Job)
 		}
 	}
+	c.operationsMutex.RUnlock()
 	return appJobs
 }
 
 // Schedule adds a Job to the Cron to be run on the given schedule.
-func (c *Cron) Schedule(schedule Schedule, job Job) {
-	entry := &Entry{
-		Schedule: schedule,
-		Job:      job,
+func (c *Cron) scheduleJob(job Job) error {
+	s, err := Parse(job.Rhythm)
+	if err != nil {
+		return err
 	}
-	if !c.running {
-		c.entriesMutex.Lock()
-		c.entries = append(c.entries, entry)
-		c.entriesMutex.Unlock()
-		return
+
+	return c.schedule(s, job)
+}
+
+// Schedule adds a Job to the Cron to be run on the given schedule.
+func (c *Cron) schedule(schedule Schedule, job Job) error {
+	partitionId := c.partitioning.CalculatePartitionId(job.Name)
+	if !c.partitioning.CheckPartitionLeader(partitionId) {
+		// It means the partitioning changed and persisted jobs are in the wrong partition now.
+		return fmt.Errorf("host does not own partition %d", partitionId)
+	}
+
+	entry := &Entry{
+		Schedule:        schedule,
+		Job:             job,
+		Prev:            time.Unix(0, 0),
+		distMutexPrefix: c.organizer.TicksPath(partitionId) + "/",
 	}
 
-	c.add <- entry
+	c.appendOperation(func(ctx context.Context) *Entry {
+		c.entries[entry.Job.Name] = entry
+		return entry
+	})
+	return nil
+}
+
+func (c *Cron) appendOperation(op func(ctx context.Context) *Entry) {
+	c.operationsMutex.Lock()
+	defer c.operationsMutex.Unlock()
+
+	c.pendingOperations = append(c.pendingOperations, op)
 }
 
 // Entries returns a snapshot of the cron entries.
 func (c *Cron) Entries() []*Entry {
 	if c.running {
-		c.entriesMutex.RLock()
-		defer c.entriesMutex.RUnlock()
-
 		c.snapshot <- nil
 		x := <-c.snapshot
 		return x
 	}
-	return c.entrySnapshot()
+
+	c.operationsMutex.RLock()
+	defer c.operationsMutex.RUnlock()
+	entries := []*Entry{}
+	for _, e := range c.entries {
+		entries = append(entries, &Entry{
+			Schedule: e.Schedule,
+			Next:     e.Next,
+			Prev:     e.Prev,
+			Job:      e.Job,
+		})
+	}
+	return entries
 }
 
 // Start the cron scheduler in its own go-routine.
 func (c *Cron) Start(ctx context.Context) {
 	c.running = true
-	go c.run(ctx)
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	c.cancel = cancel
+	go c.run(ctxWithCancel)
```

Comment:
I added this here because there is an existing Stop() function. So, the caller would need to handle the context cancelation and the Stop() call. It read off from the client's perspective.

Reply:
Let me change to remove the Stop() and depend on context cancellation only from the client side.